quickwit/quickwit-ingest/src/ingest_v2/workbench.rs

```diff
@@ -17,22 +17,156 @@
 // You should have received a copy of the GNU Affero General Public License
 // along with this program. If not, see <http://www.gnu.org/licenses/>.
 
-use std::collections::{BTreeMap, HashSet};
+use std::collections::{BTreeMap, HashMap, HashSet};
+use std::sync::{Arc, Mutex};
 
+use quickwit_common::pubsub::{EventBroker, EventSubscriptionHandle};
 use quickwit_common::rate_limited_error;
 use quickwit_proto::control_plane::{
     GetOrCreateOpenShardsFailure, GetOrCreateOpenShardsFailureReason,
 };
-use quickwit_proto::ingest::ingester::{PersistFailure, PersistFailureReason, PersistSuccess};
+use quickwit_proto::indexing::ShardPositionsUpdate;
+use quickwit_proto::ingest::ingester::{
+    PersistFailure, PersistFailureReason, PersistRequest, PersistSuccess,
+};
 use quickwit_proto::ingest::router::{
     IngestFailure, IngestFailureReason, IngestResponseV2, IngestSubrequest, IngestSuccess,
 };
 use quickwit_proto::ingest::{IngestV2Error, RateLimitingCause};
-use quickwit_proto::types::{NodeId, ShardId, SubrequestId};
-use tracing::warn;
+use quickwit_proto::types::{NodeId, Position, ShardId, SubrequestId};
+use tokio::sync::Notify;
+use tracing::{error, warn};
 
 use super::router::PersistRequestSummary;
 
+enum PublishState {
+    Tracked,
+    AwaitingPublish(Position),
+    Published(Position),
+}
+
+#[derive(Default)]
+struct ShardPublishStates {
+    states: HashMap<ShardId, PublishState>,
+    awaiting_count: usize,
+}
+
+/// A helper for awaiting shard publish events when running in `wait_for` and
+/// `force` commit mode.
+///
+/// Registers a set of shard positions and listens to [`ShardPositionsUpdate`]
+/// events to assert when all the persisted events have been published. To
+/// ensure that no events are missed:
+/// - create the tracker before any persist requests is sent
+/// - call `register_requested_shards` before each persist request to ensure that
+/// the associated publish events are recorded
+/// - call `track_persisted_shard_position` after each successful persist subrequests
+struct PublishTracker {
+    state: Arc<Mutex<ShardPublishStates>>,
+    // sync::notify instead of sync::oneshot because we don't want to store the permit
+    publish_complete: Arc<Notify>,
+    _publish_listen_handle: EventSubscriptionHandle,
+}
+
+impl PublishTracker {
+    fn new(event_tracker: EventBroker) -> Self {
+        let state = Arc::new(Mutex::new(ShardPublishStates::default()));
+        let state_clone = state.clone();
+        let publish_complete = Arc::new(Notify::new());
+        let publish_complete_notifier = publish_complete.clone();
+        let _publish_listen_handle =
+            event_tracker.subscribe(move |update: ShardPositionsUpdate| {
+                let mut state_handle = state_clone.lock().unwrap();
+                for (updated_shard_id, updated_position) in &update.updated_shard_positions {
+                    if let Some(publish_state) = state_handle.states.get_mut(updated_shard_id) {
```

Comment:
I think this code should be in a method of `ShardPublishStates`?

Reply:
why name it `state_handle`? I would suggest `shard_publish_states`, no?

Reply:
I refactored the whole thing to a separate module for more clarity